### PR TITLE
feat: add url redirect for stack report ui

### PR DESCRIPTION
**Pull request by deepak1725**
This PR adds API for URL Redirect to Stack Report UI
- Part of  [Add Report URL in CLI](https://issues.redhat.com/browse/APPAI-1702 )

### CODE REVIEW

File: bayesian/api/api_v2.py
Comment (arajkumar) [inline bayesian/api/api_v2.py:312]: Can't we extract crda_key from stack_id from the RDS?

**Review by prashbnair — CHANGES_REQUESTED** [review]

File: bayesian/api/api_v2.py
Comment (deepak1725) [reply bayesian/api/api_v2.py:312]: Nope, we need a user reference even for RDS call

Comment (deepak1725) [reply bayesian/api/api_v2.py:312]: sure, do you want user_id or crda_key, actually in cli we are calling it as `crda_key`

Comment (arajkumar) [reply bayesian/api/api_v2.py]: why do you need user reference for rds access? just stack id is enough, thats how SA GET works.

Comment (arajkumar) [reply bayesian/api/api_v2.py]: every stack id associated with user id.

Comment (arajkumar) [reply bayesian/api/api_v2.py]: look at https://github.com/fabric8-analytics/fabric8-analytics-server/blob/44ab7d7ac248aeb2e14ae7f6fbc3a57e5e8e2ae6/bayesian/utils.py#L151.
it just needs request id.

Comment (deepak1725) [reply bayesian/api/api_v2.py]: I don't feel we need a DB call for this purpose at all. 
All DB calls add to monetary costs which in this case could be avoided.

Comment (arajkumar) [reply bayesian/api/api_v2.py]: The idea here is to make the url more concise. @prashbnair WDTY?

Comment (prashbnair) [reply bayesian/api/api_v2.py]: I am fine with the current URL.

**Review by prashbnair — APPROVED** [review]

File: bayesian/default_config.py
Comment (arajkumar) [inline bayesian/default_config.py:65]: Please add corresponding entries in template.yaml

File: bayesian/api/api_v2.py
Comment (arajkumar) [inline bayesian/api/api_v2.py:308]: Is it okay to print and share the uuid?

File: bayesian/api/api_v2.py
Comment (arajkumar) [inline bayesian/api/api_v2.py:322]: Can this be a debug print? we already have access logs enabled from gunicorn

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: bayesian/api/api_v2.py
Comment (deepak1725) [reply bayesian/api/api_v2.py:322]: No, I don't think so, This logs redirect URL, which I feel is not logged by Gunicorn

File: bayesian/default_config.py
Comment (deepak1725) [reply bayesian/default_config.py:65]: Ack

File: bayesian/api/api_v2.py
Comment (arajkumar) [reply bayesian/api/api_v2.py:322]: I understand that, what I meant here is that we have minimal info print already in the form access logs for request.

File: bayesian/api/api_v2.py
Comment (arajkumar) [reply bayesian/api/api_v2.py:308]: I feel uuid is a sensitive value., shouldn't be shared as part of this.

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: openshift/template.yaml
Comment (deepak1725) [reply openshift/template.yaml:194]: we are using default values in `default_config`, what's the difference..? have we adopted any new standards..?

File: bayesian/api/api_v2.py
Comment (deepak1725) [reply bayesian/api/api_v2.py:308]: We can mask it, base64 encoded or whatever

Comment (prashbnair) [reply bayesian/api/api_v2.py:308]: I think sharing UUID is not an issue. It is required when you are using CI so why should it be different here?

File: bayesian/api/api_v2.py
Comment (deepak1725) [reply bayesian/api/api_v2.py:322]: @arajkumar This is a really helpful log, I don't think we should skip it.. It logs which URL is built by combining a number of variables.

File: bayesian/default_config.py
Comment (deepak1725) [reply bayesian/default_config.py:62]: @arajkumar That's what we are using for API authentication 
https://github.com/fabric8-analytics/fabric8-analytics-auth/blob/master/fabric8a_auth/auth.py#L101

Comment (arajkumar) [reply bayesian/default_config.py:62]: @deepak1725 thats what my point is. Stack report UI needs 3scale user key, not the pre shared key(which is private) and used between 3scale gateway and api-server.

external --> 3scale-gateway --> api-server

external to 3scale-gateway: 3scale user key(e.g. prod: 9e7da76708fe374d8c10fa752e72989f, staging: 3e42fa66f65124e6b1266a23431e3d08)
3scale-gateway to api-server: PSK, private(not to be shared)

File: bayesian/api/api_v2.py
Comment (arajkumar) [reply bayesian/api/api_v2.py:308]: In CI, user/admin takes care copying it to CI specific secrets.

However when we display it over cli, user can share it with external party which is not desirable to know the UUID.

File: openshift/template.yaml
Comment (arajkumar) [reply openshift/template.yaml:194]: it is not new standard or anything, instead of having a text "Stack Report UI (will be of Prod, if empty)", express the default value over here itself.

Comment (deepak1725) [reply openshift/template.yaml]: Ack

File: bayesian/api/api_v2.py
Comment (deepak1725) [reply bayesian/api/api_v2.py:308]: We decided to remove UUID from URL, and handle it in API.

File: bayesian/api/api_v2.py
Comment (arajkumar) [inline bayesian/api/api_v2.py:329]: Use consistent naming. It should be called as  THREESCALE_USER_KEY

File: bayesian/api/api_v2.py
Comment (deepak1725) [reply bayesian/api/api_v2.py]: Update: RDS query to fetch crda_key corresponding to particular stack _id will be made in Redirect URL function only.

File: bayesian/api/api_v2.py
Comment (deepak1725) [reply bayesian/api/api_v2.py]: Ack

File: bayesian/api/api_v2.py
Comment (arajkumar) [inline bayesian/api/api_v2.py:309]: Didn't we decided to do the RDS access on SA GET request when UUID is not present?

Comment (deepak1725) [reply bayesian/api/api_v2.py:309]: No, Prashant and I re-thought that and concluded on "RDS call in the Redirect URL" Approach would be a cleaner one.

Comment (prashbnair) [reply bayesian/api/api_v2.py:309]: The auth decorator expects a UUID , it is better to be consistent than to add hacks to work around that.

Comment (arajkumar) [reply bayesian/api/api_v2.py:309]: @prashbnair , the existing decorator can be kept as is. Without uuid, it just skips RDS call and sets the user type as unregistered. All we need to do is add one more check(the same one which @deepak1725 had put into the redir endpoint) in the SA GET which changes the user registration status after obtaining the UUID. It would be simple too and RDS access will be under 3scale which was raised as a concern during our last discussion.

Comment (deepak1725) [reply bayesian/api/api_v2.py:309]: @arajkumar Stack Report UI doesn't accept empty `uuid/crda_key`

Comment (prashbnair) [reply bayesian/api/api_v2.py:309]: @arajkumar It does not make for the user registration status to handled in different places, the auth filter does it.  That was the main reason for doing it this way. 
We need to ensure that there cannot be any SQL injection and we should be fine.

File: bayesian/api/api_v2.py
Comment (arajkumar) [inline bayesian/api/api_v2.py:58]: nit
```suggestion
from bayesian.default_config import (THREESCALE_USER_KEY,
                                                                BAYESIAN_API_HOSTNAME, 
                                                                STACK_REPORT_UI_HOSTNAME)
```

**Review by arajkumar — APPROVED** [review]
lgtm, with a nit.
please wait for @prashbnair approval.

File: bayesian/api/api_v2.py
Comment (deepak1725) [reply bayesian/api/api_v2.py]: Ack

**Pull request merged**
